azure/services/scalesetvms/scalesetvms.go

```diff
@@ -73,211 +73,95 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, log, done := tele.StartSpanWithLogger(ctx, "scalesetvms.Service.Reconcile")
 	defer done()
 
-	var (
-		resourceGroup = s.Scope.ResourceGroup()
-		vmssName      = s.Scope.ScaleSetName()
-		instanceID    = s.Scope.InstanceID()
-		providerID    = s.Scope.ProviderID()
-		isFlex        = s.Scope.OrchestrationMode() == infrav1.FlexibleOrchestrationMode
-	)
+	spec := s.Scope.ScaleSetVMSpec()
+	scaleSetVMSpec, ok := spec.(*ScaleSetVMSpec)
+	if !ok {
+		return errors.Errorf("%T is not of type ScaleSetVMSpec", spec)
+	}
 
+	reconciler := s.Reconciler
+	var getter azure.ResourceSpecGetter = scaleSetVMSpec
+	var result interface{}
+	var err error
 	// Fetch the latest instance or VM data. AzureMachinePoolReconciler handles model mutations.
-	if isFlex {
-		resourceID := strings.TrimPrefix(providerID, azureutil.ProviderIDPrefix)
-		log.V(4).Info("VMSS is flex", "vmssName", vmssName, "providerID", providerID, "resourceID", resourceID)
-		// Using VMSS Flex, so fetch by resource ID.
-		vm, err := s.VMClient.GetByID(ctx, resourceID)
+	if scaleSetVMSpec.IsFlex {
+		log.V(4).Info("VMSS is flex", "vmssName", scaleSetVMSpec.Name, "providerID", scaleSetVMSpec.ProviderID, "resourceID", scaleSetVMSpec.ResourceID)
+		getter, err = scaleSetVMSpecToVMSpec(*scaleSetVMSpec)
```

Comment:
With the change to async, should we still be fetching the vm by resource ID if it's using Flex?

Reply:
Oh I see that it has been moved within the helper function here, got it! 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3799/files?diff=split&w=0#diff-afa3aa43d3541f80b10af246e174440ced390fc557b6a907525183d4d9dd5106R157